app/src/main/java/org/mozilla/fenix/components/history/PagedHistoryProvider.kt

```diff
@@ -4,51 +4,147 @@
+import org.mozilla.fenix.FeatureFlags
+import org.mozilla.fenix.library.history.History
 import org.mozilla.fenix.perf.runBlockingIncrement
 
 /**
- * An Interface for providing a paginated list of [VisitInfo]
+ * An Interface for providing a paginated list of [History].
  */
 interface PagedHistoryProvider {
     /**
-     * Gets a list of [VisitInfo]
+     * Gets a list of [History].
+     *
      * @param offset How much to offset the list by
      * @param numberOfItems How many items to fetch
-     * @param onComplete A callback that returns the list of [VisitInfo]
+     * @param onComplete A callback that returns the list of [History]
      */
-    fun getHistory(offset: Long, numberOfItems: Long, onComplete: (List<VisitInfo>) -> Unit)
+    fun getHistory(offset: Long, numberOfItems: Long, onComplete: (List<History>) -> Unit)
 }
 
-// A PagedList DataSource runs on a background thread automatically.
-// If we run this in our own coroutineScope it breaks the PagedList
-fun HistoryStorage.createSynchronousPagedHistoryProvider(): PagedHistoryProvider {
-    return object : PagedHistoryProvider {
-
-        override fun getHistory(
-            offset: Long,
-            numberOfItems: Long,
-            onComplete: (List<VisitInfo>) -> Unit
-        ) {
-            runBlockingIncrement {
-                val history = getVisitsPaginated(
-                    offset,
-                    numberOfItems,
-                    excludeTypes = listOf(
-                        VisitType.NOT_A_VISIT,
-                        VisitType.DOWNLOAD,
-                        VisitType.REDIRECT_TEMPORARY,
-                        VisitType.RELOAD,
-                        VisitType.EMBED,
-                        VisitType.FRAMED_LINK,
-                        VisitType.REDIRECT_PERMANENT
+/**
+ * @param historyStorage
+ */
+class DefaultPagedHistoryProvider(
+    private val historyStorage: PlacesHistoryStorage,
+    private val showHistorySearchGroups: Boolean = FeatureFlags.showHistorySearchGroups
+) : PagedHistoryProvider {
+
+    private lateinit var historyGroups: List<History.Group>
+
+    @Suppress("LongMethod")
+    override fun getHistory(
+        offset: Long,
+        numberOfItems: Long,
+        onComplete: (List<History>) -> Unit,
+    ) {
+        // A PagedList DataSource runs on a background thread automatically.
+        // If we run this in our own coroutineScope it breaks the PagedList
+        runBlockingIncrement {
+            val history: List<History>
+
+            if (showHistorySearchGroups) {
+                // We need to refetch all the history metadata if the offset resets back at 0
+                // in the case of a pull to refresh.
+                if (::historyGroups.isInitialized.not() || offset == 0L) {
+                    val historyMetadata =
+                        historyStorage.getHistoryMetadataSince(Long.MIN_VALUE)
+                            .filter { it.totalViewTime > 0 && it.key.searchTerm != null }
```

Comment:
`it.totalViewTime > 0` is what's causing the problem that you saw around 15s time diff; initial observations for documentType that were recorded alongside the visits are being filtered out by this check. and we're potentially recording the viewTime observation quite a bit later than the initial documentType observation (e.g. when a tab will be closed).

at this layer, you want to look at all the data.

(in the ui though, we shouldn't display metadata record that have a zero view time).